drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -465,29 +486,122 @@ static void counter_stm32_top_irq_handle(const struct device *dev)
 }
 
-static void counter_stm32_alarm_irq_handle(const struct device *dev, uint32_t id)
+static void counter_stm32_irq_handle(const struct device *dev, uint32_t id)
 {
 	const struct counter_stm32_config *config = dev->config;
 	struct counter_stm32_data *data = dev->data;
 	TIM_TypeDef *timer = config->timer;
 
 	struct counter_stm32_ch_data *chdata;
-	counter_alarm_callback_t cb;
 
 	atomic_and(&data->cc_int_pending, ~BIT(id));
-	disable_it[id](timer);
 
 	chdata = &config->ch_data[id];
-	cb = chdata->callback;
-	chdata->callback = NULL;
+#ifdef CONFIG_COUNTER_CAPTURE
+	/* With Counter Capture, we need to check which mode it was configured for */
+	if (LL_TIM_IC_GetActiveInput(timer, LL_TIM_CHANNEL_CH1 << (4 * id)) ==
+	    LL_TIM_ACTIVEINPUT_DIRECTTI) {
+		counter_capture_cb_t cb;
+
+		cb = chdata->capture;
+
+		/* TODO: check overcapture flag, then print error?
+		 * CC1OF is also set if at least two consecutive captures
+		 * occurred whereas the flag was not cleared
+		 */
+
+		if (cb) {
+			uint32_t cc_val = get_timer_capture[id](timer);
+			uint32_t pol = LL_TIM_IC_GetPolarity(timer, LL_TIM_CHANNEL_CH1 << (4 * id));
+			uint32_t flags = 0;
+
+			/* translate stm32 flags to zephyr flags */
+			if (pol == LL_TIM_IC_POLARITY_RISING) {
+				flags = COUNTER_CAPTURE_RISING_EDGE;
+			} else if (pol == LL_TIM_IC_POLARITY_FALLING) {
+				flags = COUNTER_CAPTURE_FALLING_EDGE;
+			} else if (pol == LL_TIM_IC_POLARITY_BOTHEDGE) {
+				flags = COUNTER_CAPTURE_BOTH_EDGES;
+			}
+
+			cb(dev, id, flags, cc_val, chdata->user_data);
+		}
+	} else {
+#endif
+		counter_alarm_callback_t cb;
+
+		/* Alarm is One-Shot, so disable the Interrupt */
+		disable_it[id](timer);
 
-	if (cb) {
-		uint32_t cc_val = get_timer_compare[id](timer);
+		cb = chdata->callback;
+		chdata->callback = NULL;
 
-		cb(dev, id, cc_val, chdata->user_data);
+		if (cb) {
+			uint32_t cc_val = get_timer_compare[id](timer);
+
+			cb(dev, id, cc_val, chdata->user_data);
+		}
+#ifdef CONFIG_COUNTER_CAPTURE
 	}
+#endif
+}
+
+#ifdef CONFIG_COUNTER_CAPTURE
+static int counter_stm32_capture_enable(const struct device *dev, uint8_t chan)
+{
+	const struct counter_stm32_config *config = dev->config;
+	TIM_TypeDef *timer = config->timer;
+
+	LL_TIM_CC_EnableChannel(timer, LL_TIM_CHANNEL_CH1 << (4 * chan));
+
+	return 0;
 }
 
+static int counter_stm32_capture_disable(const struct device *dev, uint8_t chan)
+{
+	const struct counter_stm32_config *config = dev->config;
+	TIM_TypeDef *timer = config->timer;
+
+	LL_TIM_CC_DisableChannel(timer, LL_TIM_CHANNEL_CH1 << (4 * chan));
+
+	return 0;
+}
+
+static int counter_stm32_capture_callback_set(const struct device *dev, uint8_t chan,
+					      uint32_t flags, counter_capture_cb_t cb,
+					      void *user_data)
+{
+	const struct counter_stm32_config *config = dev->config;
+	struct counter_stm32_ch_data *chdata = &config->ch_data[chan];
+	TIM_TypeDef *timer = config->timer;
+	uint32_t config_flags = LL_TIM_ACTIVEINPUT_DIRECTTI;
```

Comment:
I think it would be worth the call `disable_it[chan](timer);` here.

Reply:
it shouldn't be disabling the interrupt in the case where it's being used as a counter capture, the alarm is a one-shot alarm just disables it in that case.

Reply:
I was thinking of preventing the interrupt to triggers in between `chdata->capture` and `chdata->user_data` update right below.